File: p16637.cpp
#include<bits/stdc++.h>
using namespace std;
int N, ans = -1e9;
string input;
vector<int> numbers;
vector<char> op;
int calc(char op, int n1, int n2){
    if(op == '+') return n1 + n2;
    else if(op == '-') return n1 - n2;
    else return n1 * n2;
}
void  dfs(int depth, int value){
    if(depth == op.size()){
        ans = max(ans, value);
        return;
    }
    dfs(depth + 1, calc(op[depth], value, numbers[depth + 1]));
    if(op.size() - depth >= 2){
        int num = calc(op[depth + 1], numbers[depth + 1], numbers[depth + 2]);
        dfs(depth + 2, calc(op[depth], value, num));
    }
}
int main(){
    cin >> N;
    cin >> input;
    for(int i = 0; i < N; i++){
        if(i % 2) op.push_back(input[i]);
        else numbers.push_back(input[i] - '0');
    }
    dfs(0, numbers[0]);
    cout << ans << "\n";
}

File: p16637.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class p16637 {
    public static int N, res = Integer.MIN_VALUE;
    public static List<Integer> numbers = new ArrayList<>();
    public static List<Character> operators = new ArrayList<>();

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());

        String input = br.readLine();

        for(int i = 0; i < N; i++){
            char c = input.charAt(i);

            if(i % 2 == 0) numbers.add(c - '0');
            else operators.add(c);
        }
        
        dfs(numbers.get(0), 0);

        System.out.println(res);
    }

    public static void dfs(int val, int idx){
        if(idx == N / 2){
            res = Math.max(res, val);
            return;
        }

        int nVal = calc(val, numbers.get(idx + 1), operators.get(idx));
        dfs(nVal, idx + 1);

        if(idx + 1 < N / 2){
            nVal = calc(numbers.get(idx + 1), numbers.get(idx + 2), operators.get(idx + 1));
            nVal = calc(val, nVal, operators.get(idx));
            dfs(nVal, idx + 2);
        }
    }

    public static int calc(int a, int b, char op){
        if(op == '+') return a + b;
        else if(op == '-') return a - b;
        else return a * b;
    }
}